DSharpPlus.Commands/CommandExecutor.cs

```diff
@@ -0,0 +1,137 @@
+namespace DSharpPlus.Commands;
+
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Linq;
+using System.Reflection;
+using System.Runtime.ExceptionServices;
+using System.Threading;
+using System.Threading.Tasks;
+using DSharpPlus.Commands.Commands;
+using DSharpPlus.Commands.ContextChecks;
+using DSharpPlus.Commands.EventArgs;
+using DSharpPlus.Commands.Exceptions;
+using DSharpPlus.Entities;
+using Microsoft.Extensions.DependencyInjection;
+
+public sealed class CommandExecutor
+{
+    private readonly ConcurrentDictionary<Guid, Task> _tasks = new();
+
+    /// <summary>
+    /// Executes a command asynchronously.
+    /// </summary>
+    /// <param name="context">The context of the command.</param>
+    /// <param name="block">Whether or not to block the current thread until the command has finished executing.</param>
+    /// <param name="cancellationToken">The cancellation token to use.</param>
+    /// <returns>A <see cref="ValueTask"/> representing the asynchronous operation. The value will be a <see cref="Optional{T}"/> of <see cref="bool"/>. <see cref="Optional{T}.HasValue"/> will be <see langword="true"/> if the command was executed successfully, <see langword="false"/> if the command was not executed successfully, and <see langword="null"/> if the command threw an exception.</returns>
+    public async Task ExecuteAsync(CommandContext context, bool block = false, CancellationToken cancellationToken = default)
+    {
+        Guid guid = Guid.NewGuid();
+        Task task = Task.Run(() => WorkerAsync(context), cancellationToken);
+        if (!block)
+        {
+            task = task.ContinueWith(_ => this._tasks.TryRemove(guid, out Task? _));
+        }
+
+        this._tasks.TryAdd(guid, task);
+        if (block)
+        {
+            await this._tasks[guid];
+            this._tasks.TryRemove(guid, out Task? _);
+        }
+    }
+
+    private static async ValueTask WorkerAsync(CommandContext context)
+    {
+        List<ContextCheckAttribute> checks = new(context.Command.Attributes.OfType<ContextCheckAttribute>());
+        Command? parent = context.Command.Parent;
+        while (parent is not null)
+        {
+            checks.AddRange(parent.Attributes.OfType<ContextCheckAttribute>());
+            parent = parent.Parent;
+        }
+
+        if (checks.Count != 0)
+        {
+            ContextCheckAttribute check = null!;
+            try
+            {
+                // Reverse foreach so we execute the top-most command's checks first.
+                for (int i = checks.Count - 1; i >= 0; i--)
```

Comment:
This... I have mixed feelings on. On one hand, yes it would be helpful for the user to know that they're missing X amount of required context checks. On the other hand, executing each check could cause unwanted behavior, such as the bot's internal command ratelimiting system or something similar. Though I suppose that does rely on executing the checks in order, which isn't something I want to explicitly support.